src/keyword_vpd_parser.cpp

```diff
@@ -0,0 +1,146 @@
+#include "keyword_vpd_parser.hpp"
+
+#include "constants.hpp"
+#include "exceptions.hpp"
+
+#include <iostream>
+#include <numeric>
+#include <string>
+
+namespace vpd
+{
+
+types::VPDMapVariant KeywordVpdParser::parse()
+{
+    try
+    {
+        m_vpdIterator = m_keywordVpdVector.begin();
+
+        if (*m_vpdIterator != constants::KW_VPD_START_TAG)
+        {
+            throw(
+                DataException("Invalid Large resource type Identifier String"));
+        }
+
+        checkNextBytesValidity(sizeof(constants::KW_VPD_START_TAG));
+        std::advance(m_vpdIterator, sizeof(constants::KW_VPD_START_TAG));
+
+        uint16_t dataSize = getKwDataSize();
+
+        checkNextBytesValidity(constants::TWO_BYTES + dataSize);
+        std::advance(m_vpdIterator, constants::TWO_BYTES + dataSize);
+
+        // Check for invalid vendor defined large resource type
+        if (*m_vpdIterator != constants::KW_VPD_PAIR_START_TAG)
+        {
+            if (*m_vpdIterator != constants::ALT_KW_VPD_PAIR_START_TAG)
+            {
+                throw(DataException("Invalid Keyword Vpd Start Tag"));
+            }
+        }
+        auto kwValMap = populateVpdMap();
+
+        // Do these validations before returning parsed data.
+        // Check for small resource type end tag
+        if (*m_vpdIterator != constants::KW_VAL_PAIR_END_TAG)
+        {
+            throw(DataException("Invalid Small resource type End"));
+        }
+
+        validateChecksum();
+
+        // Check vpd end Tag.
+        if (*m_vpdIterator != constants::KW_VPD_END_TAG)
+        {
+            throw(DataException("Invalid Small resource type."));
+        }
+    }
+    catch (const DataException& ex)
+    {
+        logging::logMessage(ex.what());
+
+        /*TODO: uncomment when PEL code goes in */
+
+        /*std::string errMsg = std::string{ex.what()};
+
+        inventory::PelAdditionalData additionalData{};
+        additionalData.emplace("DESCRIPTION", errMsg);
+        additionalData.emplace("CALLOUT_INVENTORY_PATH", inventoryPath);
+        createPEL(additionalData, PelSeverity::WARNING,
+                  errIntfForInvalidVPD, nullptr);*/
+    }
+    return kwValMap;
```

Comment:
Don't return the map in case of exception. re-throw the exception back to the caller from catch and let the caller decide if PEL is requried or not.

Reply:
okay